moveit_ros/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h

```diff
@@ -32,32 +32,107 @@
  *  POSSIBILITY OF SUCH DAMAGE.
  *********************************************************************/
 
-/* Author: Ioan Sucan */
+/* Author: Ioan Sucan, Sebastian Jahr
+   Description: Implementation of a MoveIt planning pipeline composed of a planner plugin and request and response
+   adapter plugins.
+*/
 
 #pragma once
 
 #include <atomic>
 
 #include <moveit/planning_interface/planning_interface.h>
-#include <moveit/planning_request_adapter/planning_request_adapter.h>
+#include <moveit/planning_interface/planning_request_adapter.h>
+#include <moveit/planning_interface/planning_response_adapter.h>
 #include <pluginlib/class_loader.hpp>
 #include <rclcpp/rclcpp.hpp>
 #include <moveit_msgs/msg/display_trajectory.hpp>
 #include <visualization_msgs/msg/marker_array.hpp>
-
 #include <memory>
-
 #include <moveit_planning_pipeline_export.h>
+#include <planning_pipeline_parameters.hpp>
 
-/** \brief Planning pipeline */
 namespace planning_pipeline
 {
-/** \brief This class facilitates loading planning plugins and
-    planning request adapted plugins.  and allows calling
-    planning_interface::PlanningContext::solve() from a loaded
-    planning plugin and the
-    planning_request_adapter::PlanningRequestAdapter plugins, in the
-    specified order. */
+/**
+ * \brief Helper function template to apply all adapters of the given vector to the motion planning problem or solution
+ *
+ * \tparam TAdapter Adapter type, PlanningRequestAdapter or PlanningResponseAdapter
+ * \param adapter_vector Adapters to be applied to request or response
+ * \param planning_scene Planning context
+ * \param req Motion plan request that describes the planning problem
+ * \param res Motion plan response to which the solution is written
+ * \param logger ROS logger
+ * \return true if all adapters are applied successfully
+ */
+template <class TAdapter>
+[[nodiscard]] bool callAdapterChain(const std::vector<TAdapter>& adapter_vector,
+                                    const planning_scene::PlanningSceneConstPtr& planning_scene,
+                                    planning_interface::MotionPlanRequest& req,
+                                    planning_interface::MotionPlanResponse& res, const rclcpp::Logger& logger)
+{
+  for (const auto& adapter : adapter_vector)
+  {
+    assert(adapter);
+    RCLCPP_INFO(logger, "Calling PlanningAdapter '%s'", adapter->getDescription().c_str());
+    // If adapter does not succeed, break chain and return false
+    if (!adapter->adapt(planning_scene, req, res))
```

Comment:
Here, we should keep track of the individual changes and errors as we go, allowing the caller to introspect what's going wrong where.